technical-docs/designs/infra/rollback.md

```diff
@@ -0,0 +1,218 @@
+# Rollback Implementation using Bedrock CLI
+
+| Revision | Date         | Author         | Remarks       |
+| -------: | ------------ | -------------- | ------------- |
+|      0.1 | Apr-20, 2020 | Nathaniel Rose | Initial Draft |
+
+## 1. Overview
+
+It is highly desireable to have a fully-automated Continuous Delivery pipeline
+in place to assist with testing and validation of source code changes made to
+the application services and supporting infrastructure. Without such a system,
+it is difficult to have any degree of confidence that a new deployment will
+succeed without introducing new regressions. However, high velocity of changes
+with distributed teams may eventually lead to the scenario in which a production
+environment transitions into a failed state due to an improperly tested feature
+added.
+
+This document outlines multiple solutions and approaches that seeks to provide a
+rollback mechanisms to ensure the successful deployment of terraform
+infrastructure, automated reverted features in a release pipeline, and system
+alerting to notify key stakeholders and operations of system activity.
+
+Rolling back services which includes a relational database can be very
+difficult. Traditionally this is done by restoring from a backup, performed
+prior to the upgrade. In more modern approach is to integrate a database
+migration tool into your application's installation procedure.
+
+Components of this design are based on the learnings from:
+
+- [Microsoft Bedrock Kubernetes Workflow Project](github.com/microsoft/bedrock)
+- [Bedrock-CLI Infrastructure Generation Pipeline](https://github.com/microsoft/bedrock-cli/blob/master/guides/infra/spk-infra-generation-pipeline.md)
+- [Terraform Recommended Practices](https://www.terraform.io/docs/cloud/guides/recommended-practices/)
+
+### User Stories Addressed:
+
+1. [Guidance on Rolling Back with SPK infra](https://github.com/microsoft/bedrock/issues/1036)
+
+## 2. Out of Scope
+
+This design document solely covers strategic approaches to revert terraform
+changes to a deployed GitOps environment in Azure DevOps after changes made have
+caused state failures. This design leverages the use of terraform state files,
+Azure storage accounts and commit hashes to build an audit history. It does not
+address the following:
+
+- Redeployment of stateful services in your environment
+- Complications arise when incompatible database changes during failed apply
+- Disaster recovery operability in respective to a fail production environment
+
+## 3. Design Details
+
+Rollback can be triggered in a separate AzDO Pipeline upon failure to build an
```

Comment:
In additional to my comment about this doc just being a part of `guides/*`, it would be useful to simply show a YAML file that represents the rollback release image behind. In general I would lean towards YAML representations of pipelines where possible.

Reply:
> YAML is not supported for release definitions. You are able to output a json
> definition for your release to apply to other projects.
> https://dev.azure.com/mseng/AzureDevOpsRoadmap/_workitems/edit/1221170/